### DISCUSSION TITLE
Automatically create IP, prefix, devices, etc. for new tenants in the same tenants group

### DISCUSSION THREAD

**Discussion by oturcot** [netbox-community/netbox#5739] [Q&A] [answered]
Hi Everyone,
I was looking at netbox to see if it could be the right IPAM for my company. Everything is perfect but there is one thing I can't seem to find.
Let me first explain our network setup.

Client A has ~500 stores. Each store has the same networking equipment (same rack, router model, same config in the router, every port configuration is the same, etc.) and all stores are connected together via multiple VPN Tunnels. Each time a new store open, we simply give them the next available subnet for their LAN, and the next available IP for each VPN Tunnel.

If I understand correctly how nexbox work, the tenants' group would be our client A, and each store would be a tenant.

Is there a way to automate the tenant creation process? So each time a new store open, I manually enter the name, street address, etc. and netbox automatically add the rack, the device, the cable, the interface, and more importantly, choses the next available IPs for the LAN and the VPN tunnels? (Is netbox already able to do that, or should I create a script and we add the new store via command line (I already did many script in shell, XML and DOS, so I could learn Python if it's necessary. ))

Thanks,
**Olivier**

**Comment by ziggekatten** [comment, marked as answer]
I would do it like this (if  all stores belongs to your Client A:

- Client A would be a Tenant
- For each store you would have a site defined and assigned to the Client A Tenant
- For create all the objects for a new store, I would create a Netbox Custom Script to do the automation. See: https://netbox.readthedocs.io/en/stable/additional-features/custom-scripts/

**Reply by oturcot** [reply]
Thanks,
Do you think it would be possible to automatically assign the IP addresses with the script?

**Reply by oturcot** [reply]
Could I create a prefix manually with all the IP addresses for the VPN, and the script take the next available IP in the prefix, and for assigning the LAN subnet, the script could also create a prefix for it, in a larger prefix?

Exemple : 
Prefix 172.16.0.0/12 for the VPN, each time I run the script for a new store, it would assign IP 172.16.0.1, then 172.16.0.2, then  172.16.0.3.
Prefix 10.0.0.0/8 for all LAN prefix, each time I run the script for a new store, it would create and assign 10.0.0.0/24, then 10.0.1.0/24, then 10.0.2.0/24.

**Reply by candlerb** [reply]
Yes, Prefix has methods [get_first_available_ip](https://github.com/netbox-community/netbox/blob/develop/netbox/ipam/models.py#L588) and [get_first_available_prefix](https://github.com/netbox-community/netbox/blob/develop/netbox/ipam/models.py#L579) - although if there were two users doing this at the same time, they'd both get the same one.

**Reply by oturcot** [reply]
By two users doing this at the same time, you mean two people running the script at the same time? Once the script is finished, we could safely run it again?

And thanks again to both of you :)

**Reply by candlerb** [reply]
Yes.  The script will pick the next IP (or prefix), and then create that IP/Prefix in the database.  If two users run the script at the same time, then they'll both try to create the same IP/Prefix - and one of them *should* fall foul of the IP uniqueness checking in Netbox and retry. But the next time you run the script, you'll get the next available IP/Prefix.

**Reply by oturcot** [reply]
Thanks :D
